### PR TITLE
Ancestry proof

### LINKED ISSUE #1441: Ancestry / prefix proofs for MMRs
For proving a BEEFY voter/signatory equivocation against a finalized signed commitment, the equivocation proofs of https://github.com/paritytech/polkadot-sdk/pull/1329 currently include headers that the verifier calculates hashes of and compares with on-chain `<frame_system::Pallet<T>>::block_hash(block_number)`. Since this mapping has a horizon of 4096 blocks, equivocations predating this horizon cannot be slashed. While this thwarts attacks assuming a 2/3rds of validators honestly participate in BEEFY finalization and at least one honest relayer can update the beefy light client at least once every 4096 blocks, this is clearly only a temporary solution. This verification is to be complemented by mmr ancestry/pre-fix proofs for equivocations older than 4096 blocks.

Concretely given a current mmr root `r`, prove that another mmr root `r'` is its ancestor.

The nervos library used atm has recently implemented a flavor of ancestry proofs (https://github.com/nervosnetwork/merkle-mountain-range/pull/35), but this requires that all the leaves `l` added to the mmr since `r'` are included in the verification call (the current implementation creates a membership proof for leaves `l`, and the proof items happen to be the peaks of `r'`). This ends up being an expensive call (`O(n_leaves_since_r')`): we might then just as well provide a headerchain back to the block including `r'`.

Efficient MMR ancestry/prefix proofs require proving membership of nodes, whereas the nervos library can currently prove leaves only (both a hardcoded assumption, as well as changes required to the proof structure).
A PoC of membership proofs for nodes can be found [here](https://github.com/Lederstrumpf/merkle-mountain-range/tree/linearize-proof).

```[tasklist]
### Tasks
- [x] Rebase PoC atop mmr library version currently used (0.5.2)
- [x] Implement ancestry proof PoC (https://github.com/Lederstrumpf/merkle-mountain-range/tree/linearize-proof-rebase-v0.5.1)
- [x] Integrate PoC into https://github.com/paritytech/polkadot-sdk/pull/1903
- [ ] Optimize PoC (in particular, proof structure can likely be simpler for proving strictly former peaks)
- [ ] Upstream to https://github.com/nervosnetwork/merkle-mountain-range/
```

### PR DESCRIPTION
**Pull request by Lederstrumpf**
# Description

Implements ancestry proofs for MMRs:

For a given MMR with size `n` and root `r`, an ancestry proof proves that some other root `r'` was the root of the MMR when it had size `n' < n`. If this is the case, we say root `r'` *ancestors* root `r`.

These ancestry proofs are used in https://github.com/paritytech/polkadot-sdk/pull/1903 to prove a correct ancestor of the current mmr root to the runtime.

Closes https://github.com/paritytech/polkadot-sdk/issues/1441

## Implementation notes
While root `r'` is not a node in the `mmr(n)`, all of its associated peaks `p': peaks(n')` are, see below illustration:
Ancestry proof - page 3 (2)

Hence, to prove `r'` ancestors `r`, we can:
1. prove membership of all `p' ∈ peaks(n')` in mmr(n)
2. verify that `calculate_root(peaks(n')) = r'`

To perform 1. we adjust the proof structure to also permit membership proofs for nodes, not just strictly leaves.
To support this the proof items also include the position of the nodes:
``` rust
pub struct NodeMerkleProof<T, M> {
    mmr_size: u64,
    proof: Vec<(u64, T)>,
    merge: PhantomData<M>,
}
```
and the proof verification is adjusted to support this.
While this was previously performed inline for `mmr::MerkleProof` (https://github.com/Lederstrumpf/merkle-mountain-range/tree/linearize-proof) and therefore also affected standard leaf proofs, this PR employs dedicated node merkle proofs implemented in `ancestry_proof::NodeMerkleProof`: https://github.com/lederstrumpf/merkle-mountain-range/tree/14edd7ceecccfec3118114f988d7dc9ef925ac20/src/ancestry_proof.rs.
While this leads to greater code duplication than adjusting `mmr::MerkleProof` to support proving node membership, they are (currently) less performant than the unadjusted leaf proofs and also unaudited. Hence, having a dedicated implementation avoids having a reaudit on areas of the code that use standard membership proofs.

### CODE REVIEW

**Review by acatangiu — APPROVED** [review]

**Review by serban300 — APPROVED** [review]

File: src/ancestry_proof.rs [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+                else if let Some(&(front_pos, ..)) = queue.front() {
+                    if height > 0 && is_descendant_pos(sib_pos, front_pos) {
+                        queue.push_back((pos, item, height));
+                        continue;
+                    } else {
+                        return Err(Error::CorruptedProof);
+                    }
+                } else {
+                    return Err(Error::CorruptedProof);
+                };
```
Comment (serban300) [inline src/ancestry_proof.rs:206]: Is this needed ? Do you have an example input where this branch would be reached ?

File: src/ancestry_proof.rs [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+}
+
+fn take_while_vec<T, P: Fn(&T) -> bool>(v: &mut Vec<T>, p: P) -> Vec<T> {
+    for i in 0..v.len() {
+        if !p(&v[i]) {
+            return v.drain(..i).collect();
+        }
+    }
+    v.drain(..).collect()
+}
```
Comment (serban300) [inline src/ancestry_proof.rs:342]: Nit: I would just use `crate::mmr::take_while_vec()` instead of duplicating it.

File: src/ancestry_proof.rs [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+pub fn bagging_peaks_hashes<T, M: Merge<Item = T>>(mut peaks_hashes: Vec<T>) -> Result<T> {
+    // bagging peaks
+    // bagging from right to left via hash(right, left).
+    while peaks_hashes.len() > 1 {
+        let right_peak = peaks_hashes.pop().expect("pop");
+        let left_peak = peaks_hashes.pop().expect("pop");
+        peaks_hashes.push(M::merge_peaks(&right_peak, &left_peak)?);
+    }
+    peaks_hashes.pop().ok_or(Error::CorruptedProof)
+}
```
Comment (serban300) [inline src/ancestry_proof.rs:315]: Nit: I would just use `crate::mmr::bagging_peaks_hashes()` instead of duplicating it.

**Review by serban300 — COMMENTED** [review]
Took a more in-depth look. Overall the approach seems sane. Just opened a PR with some small improvements here: https://github.com/paritytech/merkle-mountain-range/pull/2

@Lederstrumpf could you take a look please ?

**Pull request merged**